flink-test-utils-parent/flink-connector-testing/src/main/java/org/apache/flink/connectors/test/common/external/DefaultContainerizedExternalSystem.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connectors.test.common.external;
+
+import org.apache.flink.annotation.Experimental;
+import org.apache.flink.connectors.test.common.TestResource;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.GenericContainer;
+
+/**
+ * Default implementation of external system based on container.
+ *
+ * @param <C> Type of underlying container
+ */
+@Experimental
+public class DefaultContainerizedExternalSystem<C extends GenericContainer<C>>
+        implements TestResource {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(DefaultContainerizedExternalSystem.class);
+
+    private final GenericContainer<C> container;
+
+    /**
+     * Get a builder for {@link DefaultContainerizedExternalSystem}.
+     *
+     * @param <C> Type of underlying container
+     * @return An instance of builder
+     */
+    public static <C extends GenericContainer<C>> Builder<C> builder() {
+        return new Builder<>();
+    }
+
+    private DefaultContainerizedExternalSystem(GenericContainer<C> container) {
+        this.container = container;
+    }
+
+    @Override
+    public void startUp() throws Exception {
+        if (container.isRunning()) {
+            return;
+        }
+        container.start();
+    }
+
+    @Override
+    public void tearDown() throws Exception {
+        if (!container.isRunning()) {
+            return;
+        }
+        container.stop();
+    }
+
+    public GenericContainer<C> getContainer() {
+        return this.container;
+    }
+
+    /**
+     * Builder for {@link DefaultContainerizedExternalSystem}.
+     *
+     * @param <C> Type of underlying container
+     */
+    public static class Builder<C extends GenericContainer<C>> {
+
+        private GenericContainer<C> container;
+        private GenericContainer<?> flinkContainer;
+
+        public Builder<C> fromContainer(GenericContainer<C> container) {
+            this.container = container;
+            return this;
+        }
```

Comment:
```suggestion
        public <T extends GenericContainer<T>> Builder<T> fromContainer(
                GenericContainer<T> container) {
            @SuppressWarnings("unchecked")
            Builder<T> self = (Builder<T>) this;
            self.container = container;
            return self;
        }
```

This allows us to define kafka without `<KafkaContainer>builder`

```
    DefaultContainerizedExternalSystem<KafkaContainer> kafka =
            DefaultContainerizedExternalSystem.builder()
                    .fromContainer(
                            new KafkaContainer(DockerImageName.parse(KAFKA_IMAGE_NAME))
                                    .withNetworkAliases(KAFKA_HOSTNAME)) // captures the type here!
                    .bindWithFlinkContainer(flink.getFlinkContainer())
                    .build();
```